blockchain/message_pool/src/msgpool.rs

```diff
@@ -0,0 +1,951 @@
+        self.cs.subscribe()
+    }
+
+    fn get_heaviest_tipset(&mut self) -> Option<Tipset> {
+        let ts = self.cs.heaviest_tipset()?;
+        Some(ts.as_ref().clone())
+    }
+
+    fn put_message(&self, msg: &SignedMessage) -> Result<Cid, Error> {
+        let cid = self
+            .cs
+            .db
+            .put(msg, Blake2b256)
+            .map_err(|err| Error::Other(err.to_string()))?;
+        Ok(cid)
+    }
+
+    fn state_get_actor(&self, addr: &Address, ts: &Tipset) -> Result<ActorState, Error> {
+        let state = StateTree::new_from_root(self.cs.db.as_ref(), ts.parent_state())
+            .map_err(Error::Other)?;
+        let actor = state.get_actor(addr).map_err(Error::Other)?;
+        actor.ok_or_else(|| Error::Other("No actor state".to_owned()))
+    }
+
+    fn messages_for_block(
+        &self,
+        h: &BlockHeader,
+    ) -> Result<(Vec<UnsignedMessage>, Vec<SignedMessage>), Error> {
+        chain::block_messages(self.cs.blockstore(), h).map_err(|err| err.into())
+    }
+
+    fn messages_for_tipset(&self, h: &Tipset) -> Result<Vec<UnsignedMessage>, Error> {
+        self.cs.messages_for_tipset(h).map_err(|err| err.into())
+    }
+
+    fn load_tipset(&self, tsk: &TipsetKeys) -> Result<Tipset, Error> {
+        self.cs.tipset_from_keys(tsk).map_err(|err| err.into())
+    }
+}
+
+/// This is the main MessagePool struct
+pub struct MessagePool<T: 'static> {
+    local_addrs: Vec<Address>,
+    pending: HashMap<Address, MsgSet>,
+    pub cur_tipset: Mutex<Tipset>,
+    api: Arc<Mutex<T>>,
+    pub min_gas_price: BigInt,
+    pub max_tx_pool_size: i64,
+    pub network_name: String,
+    bls_sig_cache: LruCache<Cid, Signature>,
+    sig_val_cache: LruCache<Cid, ()>,
+    local_msgs: HashMap<Vec<u8>, SignedMessage>,
+}
+
+impl<T> MessagePool<T>
+where
+    T: Provider + std::marker::Send,
+{
+    /// Create a new message pool
+    pub fn new(mut api: T, network_name: String) -> Result<Arc<Mutex<MessagePool<T>>>, Error>
+    where
+        T: Provider,
+    {
+        // LruCache sizes have been taken from the lotus implementation
+        let tipset = Mutex::new(
+            api.get_heaviest_tipset()
+                .ok_or_else(|| Error::Other("No ts in api to set as cur_tipset".to_owned()))?,
+        );
+        let bls_sig_cache = LruCache::new(40000);
+        let sig_val_cache = LruCache::new(32000);
+        let api_mutex = Arc::new(Mutex::new(api));
+        let mp = Arc::new(Mutex::new(MessagePool {
+            local_addrs: Vec::new(),
+            pending: HashMap::new(),
+            cur_tipset: tipset,
+            api: api_mutex,
+            min_gas_price: Default::default(),
+            max_tx_pool_size: 5000,
+            network_name,
+            bls_sig_cache,
+            sig_val_cache,
+            local_msgs: HashMap::new(),
+        }));
+
+        let mut mp_lock = mp.lock().map_err(|_| Error::MutexPoisonError)?;
```

Comment:
You can just unwrap here I think if youre locking a mutex... Not sure how you would recover from a Poisoned Mutex

Reply:
The issue that I was having with simply calling unwrap was if there was a MutexPoisonError, It wasn't always due to a poisioned lock. For example, in the last test in messagepool, if you remove the sleep call, there will be a MutexPoisonError due to the subscribe new heads async block using the messagepool. Adding that map_err accounted for that so there wouldn't be any crashes due to mutexes that weren't really poisoned.

Reply:
would be nice to use the async locks (from async-std) and you wouldn't have to worry about this at all, but I do see the increased dev in making functions async

Reply:
That's weird... I don't see why the mutex would error if it wasnt poisoned...

Reply:
I know, I was trying to understand why that was going on, but after adding the sleep call in the test, it went away.

Reply:
I would also consider using the mutexes provided by async_std instead of the std lib one. The issue could be coming from the fact we're mixing them

Reply:
The mutexes in the std might also block the future which is problematic